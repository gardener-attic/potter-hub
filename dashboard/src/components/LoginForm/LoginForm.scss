.LoginForm__container {
  margin: 5em auto;
  width: 500px;
}

.LoginForm > .bg-skew .bg-skew__content {
  max-width: 75%;
}
