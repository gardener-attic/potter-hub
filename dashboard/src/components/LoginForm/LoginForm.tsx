import { Location } from "history";
import * as React from "react";
import { Lock } from "react-feather";
import { Redirect } from "react-router";

import LoadingWrapper from "../../components/LoadingWrapper";
import "./LoginForm.css";

interface ILoginFormProps {
  authenticated: boolean;
  authenticating: boolean;
  authenticationError: string | undefined;
  tryToAuthenticateWithOIDC: (validateTokenOnTargetCluster: boolean) => void;
  location: Location;
  targetClusterSecretNamespace: string;
  targetClusterSecretName: string;
}

interface ILoginFormState {
  token: string;
}

class LoginForm extends React.Component<ILoginFormProps, ILoginFormState> {
  public state: ILoginFormState = { token: "" };

  public componentDidMount() {
    const validateTokenOnTargetCluster = this.uiCalledForTargetCluster()
    this.props.tryToAuthenticateWithOIDC(validateTokenOnTargetCluster);
  }

  public render() {
    if (this.props.authenticating) {
      return <LoadingWrapper />;
    }
    if (this.props.authenticated) {
      const { from } = this.props.location.state || { from: { pathname: "/" } };
      return <Redirect to={from} />;
    }
    return (
      <section className="LoginForm">
        <div className="LoginForm__container padding-v-bigger bg-skew">
          <div className="container container-tiny">
            {this.props.authenticationError && (
              <div className="alert alert-error margin-c" role="alert">
                An error occured during authentication. Troubleshooting:
                <ul>
                    <li>Check that the namespace and secret name of the target cluster in the URL are correct.</li>
                    <li>Check that the target cluster is reachable.</li>
                    <li>Check that your token is valid on the target cluster.</li>
                </ul>
              </div>
            )}
          </div>
          <div className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse">
            <div className="container">
              <h2>
                <Lock /> Login
              </h2>
              <p>
                Authentication failed.
              </p>
            </div>
          </div>
        </div>
      </section>
    );
  }

  private uiCalledForTargetCluster = () => {
    return this.props.targetClusterSecretNamespace !== "" && this.props.targetClusterSecretName !== "";
  }

}

export default LoginForm;
