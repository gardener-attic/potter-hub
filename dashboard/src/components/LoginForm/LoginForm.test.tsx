import { shallow } from "enzyme";
import { Location } from "history";
import context from "jest-plugin-context";
import * as React from "react";
import { Redirect } from "react-router";
import itBehavesLike from "../../shared/specs";

import LoginForm from "./LoginForm";

const emptyLocation: Location = {
  hash: "",
  pathname: "",
  search: "",
  state: "",
};

const defaultProps = {
  authenticate: jest.fn(),
  authenticated: false,
  authenticating: false,
  authenticationError: undefined,
  location: emptyLocation,
  tryToAuthenticateWithOIDC: jest.fn(),
};

const authenticationError = "it's a trap";

describe("componentDidMount", () => {
  it("should call tryToAuthenticateWithOIDC", () => {
    const tryToAuthenticateWithOIDC = jest.fn();
    shallow(<LoginForm {...defaultProps} tryToAuthenticateWithOIDC={tryToAuthenticateWithOIDC} />);
    expect(tryToAuthenticateWithOIDC).toHaveBeenCalled();
  });
});

context("while authenticating", () => {
  itBehavesLike("aLoadingComponent", {
    component: LoginForm,
    props: { ...defaultProps, authenticating: true },
  });
});

it("does not render the token login form", () => {
  const wrapper = shallow(<LoginForm {...defaultProps} />);
  expect(wrapper.find("input#token").exists()).toBe(false);
  expect(wrapper.find(Redirect).exists()).toBe(false);
  expect(wrapper).toMatchSnapshot();
});

it("renders a hint for access control", () => {
  const wrapper = shallow(<LoginForm {...defaultProps} />);
  expect(wrapper.html()).toMatch(
    /Authentication failed./,
  );
});

describe("redirect if authenticated", () => {
  it("redirects to / if no current location", () => {
    const wrapper = shallow(<LoginForm {...defaultProps} authenticated={true} />);
    const redirect = wrapper.find(Redirect);
    expect(redirect.props()).toEqual({ to: { pathname: "/" } });
  });

  it("redirects to previous location", () => {
    const location = Object.assign({}, emptyLocation);
    location.state = { from: "/test" };
    const wrapper = shallow(
      <LoginForm {...defaultProps} authenticated={true} location={location} />,
    );
    const redirect = wrapper.find(Redirect);
    expect(redirect.props()).toEqual({ to: "/test" });
  });
});

it("displays an error if the authentication error is passed", () => {
  const wrapper = shallow(
    <LoginForm {...defaultProps} authenticationError={authenticationError} />,
  );

  expect(wrapper.find(".alert-error").exists()).toBe(true);
  expect(wrapper).toMatchSnapshot();
});
