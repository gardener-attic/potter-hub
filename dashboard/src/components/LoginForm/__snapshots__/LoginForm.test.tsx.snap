// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays an error if the authentication error is passed 1`] = `
<section
  className="LoginForm"
>
  <div
    className="LoginForm__container padding-v-bigger bg-skew"
  >
    <div
      className="container container-tiny"
    >
      <div
        className="alert alert-error margin-c"
        role="alert"
      >
        An error occured during authentication. Troubleshooting:
        <ul>
          <li>
            Check that the namespace and secret name of the target cluster in the URL are correct.
          </li>
          <li>
            Check that the target cluster is reachable.
          </li>
          <li>
            Check that your token is valid on the target cluster.
          </li>
        </ul>
      </div>
    </div>
    <div
      className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse"
    >
      <div
        className="container"
      >
        <h2>
          <Lock
            color="currentColor"
            size="24"
          />
           Login
        </h2>
        <p>
          Authentication failed.
        </p>
      </div>
    </div>
  </div>
</section>
`;

exports[`does not render the token login form 1`] = `
<section
  className="LoginForm"
>
  <div
    className="LoginForm__container padding-v-bigger bg-skew"
  >
    <div
      className="container container-tiny"
    />
    <div
      className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse"
    >
      <div
        className="container"
      >
        <h2>
          <Lock
            color="currentColor"
            size="24"
          />
           Login
        </h2>
        <p>
          Authentication failed.
        </p>
      </div>
    </div>
  </div>
</section>
`;

exports[`while authenticating loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;
